test/lib/mock-blpapi.ts

```diff
@@ -0,0 +1,272 @@
+var SERVICE_TO_SUBSCRIPTION_EVENTS_MAP: { [uri: string]: string[]; } = {
+    '//blp/mktdata':  ['MarketDataEvents'],
+    '//blp/mktvwap':  ['MarketDataEvents'],
+    '//blp/mktbar':   ['MarketBarStart', 'MarketBarUpdate', 'MarketBarEnd'],
+    '//blp/pagedata': ['PageUpdate']
+};
+
+function getServiceForSecurity(security: string): string
+{
+    var serviceRegex = /^\/\/blp\/[a-z]+/;
+    var match = serviceRegex.exec(security);
+    // XXX: note that we shoud probably capture what the default service is to use when
+    //      reading in the session options.  However, when not specified, it is
+    //      '//blp/mktdata'.
+    return match ? match[0] : '//blp/mktdata';
+}
+
+// Mock Session object
+export class Session extends events.EventEmitter implements ISession {
+    // TypeScript compiler needs this to allow "this['property-string']" type of access
+    [index: string]: any;
+
+    // PRIVATE MANIPULATORS
+    private terminateSession(): void {
+        this.emit('SessionTerminated', { data: 'Session Terminated.' });
+    }
+
+    private sendSessionStarted(): void {
+        this.emit('SessionStarted', { data: 'Session Started.' });
+    }
+
+    private sendSessionStartupFailure(): void {
+        this.emit('SessionStartupFailure', {
+            data: { reason: { description:  'Session Fail to Start.' } }
+        });
+    }
+
+    private sendServiceOpened(cid: number): void {
+        this.emit('ServiceOpened', { correlations: [{ value: cid }] });
+    }
+
+    private sendServiceOpenFailure(uri: string, cid: number): void {
+        this.emit('ServiceOpenFailure', {
+            data: { reason: { description:  uri + ' Service Fail to Open.' } },
+            correlations: [{ value: cid }]
+        });
+    }
+
+    private sendPartialRequestData(responseName: string, cid: number): void {
+        this.emit(responseName, {
+            eventType: 'PARTIAL_RESPONSE',
+            messageType: responseName,
+            correlations: [{ value: cid }],
+            data: 'TestData'
+        });
+    }
+
+    private sendFinalRequestData(responseName: string, cid: number): void {
+        this.emit(responseName, {
+            eventType: 'RESPONSE',
+            messageType: responseName,
+            correlations: [{ value: cid }],
+            data: 'FinalTestData'
+        });
+    }
+
+    private sendSubscriptionData(eventName: string, cid: number): void {
+        this.emit(eventName, {
+            correlations: [{ value: cid }],
+            data: 'TestData'
+        });
+    }
+
+    // Error throwing must be synchronous
+    private throwError(): void {
+        throw new Error('TestError');
+    }
+
+    // CREATORS
+    constructor (args: SessionOpts) {
+        super();
+
+        ipc.on('terminate-session', (): void => {
+            this.terminateSession();
+        });
+    }
+
+    // MANIPULATORS
+    start(): Session {
+        // If has preset instructions, follow it
+        if (_.has(instructions, 'start')) {
```

Comment:
The only `instruction` that cannot be set on-demand is the **throwError** command that needs to be happen synchronously.